cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -220,86 +218,79 @@ func (cm *certManagerClient) createObj(o unstructured.Unstructured) error {
 	return nil
 }
 
-// getWebhook returns the cert-manager Webhook or nil if it does not exists.
-func (cm *certManagerClient) getWebhook() (*unstructured.Unstructured, error) {
+// allObjectsExist checks if all the objects with the given name already exist
+func (cm *certManagerClient) allObjectsExist(ctx context.Context, log logr.Logger, objs []unstructured.Unstructured) (bool, error) {
 	c, err := cm.proxy.NewClient()
 	if err != nil {
-		return nil, err
+		return false, err
 	}
 
-	webhook := &unstructured.Unstructured{}
-	webhook.SetAPIVersion("apiregistration.k8s.io/v1beta1")
-	webhook.SetKind("APIService")
-	webhook.SetName("v1beta1.webhook.cert-manager.io")
+	for _, obj := range objs {
+		log.V(5).Info("Checking", logf.UnstructuredToValues(obj)...)
+		key, err := client.ObjectKeyFromObject(&obj)
+		if err != nil {
+			return false, err
+		}
 
-	key, err := client.ObjectKeyFromObject(webhook)
-	if err != nil {
-		return nil, err
-	}
 
-	err = c.Get(ctx, key, webhook)
-	if err != nil {
-		if !apierrors.IsNotFound(err) {
-			return nil, err
+		// create a DeepCopy of the object so we can perform the Get into it
+		// without modifying the stored value
+		newObj := obj.DeepCopyObject()
+		if err := c.Get(ctx, key, newObj); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+
+			return false, err
 		}
-		return nil, nil
 	}
 
-	return webhook, nil
+	return true, nil
 }
 
-// hasWebhook returns true if there is already a web-hook in the cluster
-func (cm *certManagerClient) hasWebhook() (bool, error) {
-	// Checks if the cert-manager web-hook already exists, if yes, no additional images are required
-	webhook, err := cm.getWebhook()
-	if err != nil {
-		return false, errors.Wrap(err, "failed to check if the cert-manager web-hook exists")
-	}
-	if webhook != nil {
-		return true, nil
-	}
-	return false, nil
-}
+// waitForAPIReady will attempt to create the cert-manager 'test assets' (i.e. a basic
+// Issuer and Certificate).
+// This ensures that the Kubernetes apiserver is ready to serve resources within the
+// cert-manager API group.
+func (cm *certManagerClient) waitForAPIReady(ctx context.Context, log logr.Logger) error {
```

Comment:
Just for confirmation, in this func we are creating an issurer/a certificate, but we are not waiting for the certificate to be processed as documented https://cert-manager.io/docs/installation/kubernetes/#verifying-the-installation.
Is this enough for assuming cert-manager is working?

Reply:
Another question, as far as I know cert-manager, this test can work both with v0.11.0 and v0.15.2

Reply:
> Is this enough for assuming cert-manager is working?

So it depends on how you want to handle this - you are correct, that it may not yet be ready to issue certificates as the controller could still be leader electing. However, presumably the parts of code that create the Certificate resource also then wait for those Certificates to be *Ready* before consuming the signed certificate/private key. Would you rather wait at the start for the controller to be ready, or wait until further into execution? Easy enough to change either way 😄 

> Another question, as far as I know cert-manager, this test can work both with v0.11.0 and v0.15.2

Correct, this is a far more universal test - if this stops working, it'll be because:

1) The v1alpha2 API version no longer exists in cert-manager
2) We made a big mistake in cert-manager 😅